### ISSUE TITLE
Weird results on RTD2719W firmware

### ISSUE THREAD

**Issue by c0d3z3r0** [8051Enthusiast/at51#7]
at51 libfind is one of my standard tools when reverse-engineering 8051 firmwares. When working on a firmware binary for RTD2719W, I get strange results, though.

Binary (well, actually just the first bank): 
[rtd_000000.bin.zip](https://github.com/8051Enthusiast/at51/files/14422924/rtd_000000.bin.zip)

Running libfind with Keil's C51.LIB produces good results:
```
$ at51 libfind rtd_split/rtd_000000.bin C51C.LIB
Address | Name                 | Description
0x00fe    ?C?COPY               
...
0x079d   [?C_START]             
0x0a7b   (MAIN)                 
0x42da    ?C?PLDXDATA            general pointer load from xdata
0x5be6    ?C?PLDCODE             general pointer load from code space
```

Then I built a project with a veeeery similiar codebase (and many matching functions) as C51 library using Keil UV5: [linsn_2796Code](https://github.com/weimuxing/linsn_2796Code/tree/master/source)
Resulting lib: 
[RL6193_Project.LIB.zip](https://github.com/8051Enthusiast/at51/files/14423024/RL6193_Project.LIB.zip)

Running libfind produces repeating blocks of a bunch of functions and only very few good matches: (full output: [libfind.txt](https://github.com/8051Enthusiast/at51/files/14423063/libfind.txt))

```
$ at51 libfind rtd_split/rtd_000000.bin RL6193_Project.LIB
Address | Name                 | Description
0x0006   (RTDFACTORYOSDFUNCDISABLEOSD) 
0x0006   (SCALEROSDENABLEOSD)   
0x0006   (SCALERSYNCTMDSHPDTOGGLEPROC) 
0x0006   (USERCOMMONEEPROMSAVEMODEUSERDATA) 
0x0006   (USERDDCCIHANDLER)     
0x0006   (_SCALERCOLORDCCNORMALIZEFACTORADJUST) 
0x0006   (_SCALERCOLORSIXCOLORINITIALNORMAL) 
0x0006   (_SCALERMCUUARTWRITE)  
0x0006   (_SCALEROSDSETTRANSPARENCY) 
0x0006   (_SCALEROSDWINDOWDISABLE) 
0x0016   (SCALERAUDIODIGITALAUDIOINITIAL) 
0x0016   (SCALERFRCINITIAL)     
0x0016   (USERCOMMONAUTOCONFIG) 
0x00fe   (RTDFACTORYOSDFUNCDISABLEOSD) 
0x00fe   (SCALEROSDENABLEOSD)   
0x00fe   (SCALERSYNCTMDSHPDTOGGLEPROC) 
0x00fe   (USERCOMMONEEPROMSAVEMODEUSERDATA) 
0x00fe   (USERDDCCIHANDLER)     
0x00fe   (_SCALERCOLORDCCNORMALIZEFACTORADJUST) 
...
0x5dbe    SCALERTMDSRX1TMDSVIDEODETECTEVENT  
0x5dbe    SCALERTMDSRX2TMDSVIDEODETECTEVENT  
0x5dbe    SCALERTMDSRX4TMDSVIDEODETECTEVENT  
0x5dbe    SCALERTMDSRX5TMDSVIDEODETECTEVENT  
0x5e80   (SCALERAUDIODIGITALAUDIOINITIAL) 
0x5e80   (SCALERFRCINITIAL)     
0x5e80   (USERCOMMONAUTOCONFIG) 
0x5f4c    _SCALERGLOBALCRYSTALCLKSEL  
0x5f5d    SCALERAUDIODACCLRSTATECHANGE  
0x5f7e   (SCALERPINSHAREPOWERONRESET) 
0x5fa6    GETINPUTCAPFOUNTION   
0x5fa6    SCALERCOLORPCMGETTABLEBANK  
0x5fa6    SCALERDEBUGGETDDCCIDEBUGMODE  
0x5fa6    SYSMODEGETDISPLAYMODE  
0x5fa6    USERINTERFACEGETSHARPNESSCOEFBANKNUM  
0x5fac    SCALERMCUCACHEINITIAL  
0x5fbb    SCALERDPTXHDCPTIMEOUTFORVREADYEVENT  
0x5fc3    _SCALERTIMERDELAYXMS  
...
```

Any idea, what's wrong here?

**Comment by 8051Enthusiast** [comment]
Can you retry with the current master? Normally, libfind filters segments with less than 4 known bytes, as that is too likely to match and produce garbage. However, I forgot to apply that to references. In this case, some segments of the form `LJMP $ADDREESS` got matched and produced a reference to the corresponding symbol everytime there was an `LJMP` opcode.

There will probably still be multiple function possibilities for the same address. For example:
```
0x5fa6    GETINPUTCAPFOUNTION   
0x5fa6    SCALERCOLORPCMGETTABLEBANK  
0x5fa6    SCALERDEBUGGETDDCCIDEBUGMODE  
0x5fa6    SYSMODEGETDISPLAYMODE  
0x5fa6    USERINTERFACEGETSHARPNESSCOEFBANKNUM
```
are all functions that just read a single byte from XDATA. `libfind` currently looks only at the dependencies of a function, and not at reverse dependencies, so it cannot figure that one out.

**Comment by c0d3z3r0** [comment]
Wow, thanks for looking into it that quick! :-)
Would it make sense to add a filter for too small functions, e.g. `--min-fn-length 20` (bytes). Maybe also an option for skipping multiple matches per offset?

**Comment by 8051Enthusiast** [comment]
Oh yeah, I should actually just make the current <4 bytes filter into an configurable option. Not sure about how I should handle multiple matches, since there doesn't seem to be a clear way to prefer one over the other.

**Comment by c0d3z3r0** [comment]
Ah well, what I mean is skipping an offset completely / not printing anything if there are more than n matches

**Comment by 8051Enthusiast** [comment]
Yeah that makes sense. I've tried implementing these options, and with `at51 --min-fn-length 10 rtd_000000.bin RL6193_Project.LIB` i get just
```
Address | Name                 | Description
0x5f4c    _SCALERGLOBALCRYSTALCLKSEL
0x5fac    SCALERMCUCACHEINITIAL
0x5fc3    _SCALERTIMERDELAYXMS
0x6025    _SCALERM2PLLDIV
0x6045    SCALERGLOBALIRQCONTROL
```
`libfind` probably just works better with the standard library because i think it is written in assembly, while the compiled versions might change more across compiler versions

**Comment by c0d3z3r0** [comment]
Yeah, I guess that I need to find out what UV version was used originally and see if there are better results then. I'll keep you posted.

Anyway, thank you so much for your support and your awesome work!

**Comment by c0d3z3r0** [comment]
Update: I automated building that project with a all available of uv versions (v7.10-v9.61; starting from v710 bc everything earlier is too old to build the code). at51 libfind results don't differ at all, so the problem must be something else... I'll try to get building a full binary working and check that against the built libs. Let's see if that works.

**Comment by c0d3z3r0** [comment]
Ok, got a binary now (versions 9.54-9.61 work, but the linker fails. I was able to run it from command line successfully, though... weird thing). The result looks a bit better on bank0, following banks look pretty good actually, even when checking against a lib compiled with a different uv version (7.10 in this example).  

[bin.zip](https://github.com/8051Enthusiast/at51/files/14465027/bin.zip)
[libfind_bank0.txt](https://github.com/8051Enthusiast/at51/files/14465028/libfind_bank0.txt)
[libfind_bank1.txt](https://github.com/8051Enthusiast/at51/files/14465029/libfind_bank1.txt)
[libfind_bank2.txt](https://github.com/8051Enthusiast/at51/files/14465030/libfind_bank2.txt)

I start to wonder if either a different linker was used for the dumped binary I am looking at or even if a different compiler was used :/
